edocumentation-domain/src/main/java/org/dcsa/edocumentation/domain/persistence/state/BookingStateMachineImpl.java

```diff
@@ -0,0 +1,85 @@
+package org.dcsa.edocumentation.domain.persistence.state;
+
+import lombok.Getter;
+import org.dcsa.edocumentation.domain.persistence.entity.enums.BkgDocumentStatus;
+
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.dcsa.edocumentation.domain.persistence.state.BookingStateMachineTables.BOOKING_STATE_TRANSITION_TABLE;
+import static org.dcsa.edocumentation.domain.persistence.state.BookingStateMachineTables.INITIAL_STATE;
+
+public class BookingStateMachineImpl implements BookingStateMachine {
+
+  @Getter
+  BkgDocumentStatus currentStatus;
+
+  BookingStateMachineImpl() {
+    currentStatus = INITIAL_STATE;
+  }
+
+  BookingStateMachineImpl(BkgDocumentStatus currentStatus) {
+    this.currentStatus = Objects.requireNonNull(currentStatus, "currentStatus cannot be null");
+    verifyState(currentStatus);
+  }
+
+  private void transitionTo(BkgDocumentStatus successor) {
+    verifyTransition(successor);
+    currentStatus = successor;
+  }
+
+  void verifyTransition(BkgDocumentStatus intendedSuccessor) {
+    Objects.requireNonNull(intendedSuccessor, "successor status cannot be null");
+    Set<BkgDocumentStatus> successors = BOOKING_STATE_TRANSITION_TABLE.get(this.currentStatus);
+    if (successors == null) {
+      throw new IllegalArgumentException("Unknown/Unhandled state: " + this.currentStatus.name());
+    }
+    if (!successors.contains(intendedSuccessor)) {
+      String detail = "It is a terminal state!";
+      if (!successors.isEmpty()) {
+        detail = "The following states are valid: " + successors.stream()
+          .map(BkgDocumentStatus::name)
+          .collect(Collectors.joining(", ", "[", "]"));
+      }
+      throw new IllegalArgumentException("Invalid transition: " + this.currentStatus.name() + " *CANNOT* got to "
```

Comment:
Should we throw a DCSA exception here instead?

Reply:
Ditto